### ISSUE TITLE
Support for PowerBanks

### ISSUE THREAD

**Issue by thmsndk** [thmsndk/Screeps3D#23]
bonzaiferroni#5

https://docs.screeps.com/api/#StructurePowerBank
https://docs.screeps.com/power.html#Power-Banks

![official-powerbank](https://user-images.githubusercontent.com/9084377/60403989-2fc6a480-9ba4-11e9-9bbb-4de9b7259483.gif)
Has the following description in the side panel
> This structure emits bright light and splashes of sparks.
Write more detailed tasklist
- [x] Model
- [ ] Animate taking damage
- [ ] Animate return damage
- [x] visualize capacity
- [ ] visualize decay

**Comment by thmsndk** [comment]
KittyTack on slack suggested the following as visuals
> The pBank could be a semitransparent parallelepiped with a red pulsing sphere in it,

**Comment by thmsndk** [comment]
Initial base model
![image](https://user-images.githubusercontent.com/9084377/60404315-10ca1180-9ba8-11e9-995a-0d29d5473331.png)

**Comment by thmsndk** [comment]
Added an "animated spherelike energy source" does not animate ingame / in unity though... does in blender. 🤔 😕 

![image](https://user-images.githubusercontent.com/9084377/60431263-a3e26600-9bff-11e9-8d68-76894f43628b.png)
Animation inside blender
![animated-powersource-powerbank](https://user-images.githubusercontent.com/9084377/60431452-23703500-9c00-11e9-88df-aac427686a0a.gif)

**Comment by thmsndk** [comment]
https://www.reddit.com/r/screeps/comments/6nuyx3/basic_question_about_power_banks/
>As far as I can tell from the docs, you have to attack the power bank (and deal with the reflecting damage) to access the actual power resource. But once you "crack" the Power Bank does the power resource just drop onto the floor, does it need to then be "Harvested", does the broken Power Bank act as a container? What happens?

>Its dropped on the floor in a resource pile. It decays like any other resource pile so u need creeps ready to pick it up.

**Comment by thmsndk** [comment]
Could do something like this to visualize hitpoints on ther powerBank?
Glowing cracks:
https://www.youtube.com/watch?v=UEdA5pP5Wrk
Shattering objects:
https://www.reddit.com/r/Unity3D/comments/65kcys/learn_to_realistically_shatter_game_objects_using/
https://www.youtube.com/watch?v=btagDVtChtg

**Comment by thmsndk** [comment]
When a powerBank is killed a resource containing power is dropped
![image](https://user-images.githubusercontent.com/9084377/60736565-b2aa7f00-9f57-11e9-8775-6a4f1ff0ae07.png)
this resource is rendered inside the wall and is not colored accordingly to the resource type (power)
![image](https://user-images.githubusercontent.com/9084377/60736547-a9211700-9f57-11e9-8f53-bc203db75d52.png)
![image](https://user-images.githubusercontent.com/9084377/60736615-df5e9680-9f57-11e9-849c-1e4252513f6c.png)

**Comment by thmsndk** [comment]
I have trouble figuring out if the creep is hitting a powerBank, and when the creep is getting hit from the powerBank.

Basicly we recieve the following actions on the creep
- attack
- attacked
- rangedAttack
The only thing they contain though is a location (x,y) 

We could let us inspire by the following and perhaps make a RayTrace.
```
EffectsUtility.Beam(_creep, _creep.Actions[beam.Key], beam.Value);
var endPos = PosUtility.Convert(target, origin.Room) + new Vector3(0, config.EndHeight, 0);
```

Alternatively we could lookup what items are currently on that position, will it work with creeps shooting creeps though? what if there is both a creep and a road tile, how do we determine what the creep is actually shooting at?

**Comment by ssmith151** [comment]
This is a delicate issue.  When the room is made all the objects are given IDs.  Those IDs need saved while you are subscribed to a room.  Those IDs also have room positions.  A dictionary needs to constructed to store IDs at those positions.  During the tick where something is attacked you have that action log for the attacking object to the position.  That position gets looked up to find the IDs.  Look at each ID and all of the data came in this tick to see if "hits" is in JSONobject.  Solves most scenarios.  Not the case for multiple aggressors attacking differing structures in the same space though.  The renderer does no distinguish between these either.  It might be best to make it ambiguous.

tl;dr Hits being reduced is a good way to solve most of this.

**Comment by thmsndk** [comment]
🤔 I guess we should make a new issue to handle targeting in general, I feel that the support for powerbanks are good enough or atleast on par with the official client now. i'll make a PR for it in its current state very soon unless someone do not agree :)

**Comment by Gadjung** [comment]
Decay is implemented (as per scaling particles amount)
for showing doing/returning dmg i'm not sure it's possible

**Comment by thmsndk** [comment]
@Gadjung can't we find all creeps in range 3 of it and check if one of their actions was attack/rangedattack/rma on the powerbank?  not sure we have other options, not sure on the data we recieve it is just a position, but doing the lookup might be feasable :)

**Comment by Gadjung** [comment]
one way would be to shift part of creep actions to the 'receiving end' like i've done with heal (that heal effect is played on creep that has `healed action`)
